File: netlify/functions/rating.py
import json
import re
import requests

def get_codechef_data(username):
    """
    Fetches a user's profile page and extracts the embedded rating data JSON.
    """
    url = f"https://www.codechef.com/users/{username}"
    headers = {'User-Agent': 'Mozilla/5.0'}
    
    try:
        response = requests.get(url, headers=headers)
        response.raise_for_status()
        
        match = re.search(r'jQuery\.extend\(Drupal\.settings, (\{.*\})\);', response.text)
        if not match:
            return None, "Could not find data object in page source."
            
        data = json.loads(match.group(1))
        return data, None
        
    except requests.exceptions.HTTPError:
        return None, f"User '{username}' not found or profile is private."
    except Exception as e:
        return None, f"An unexpected error occurred: {str(e)}"

def handler(event, context):
    """
    This is the main entry point for the Netlify Function.
    It processes the incoming request and returns a JSON response.
    """
    # Handle case where queryStringParameters might be None
    query_params = event.get('queryStringParameters') or {}
    username = query_params.get('username')

    # --- THIS IS THE MODIFIED LOGIC ---
    # If no username is provided (which happens when the root URL is hit),
    # return a helpful welcome message instead of an error.
    if not username:
        welcome_message = {
            "message": "Welcome to the Unofficial CodeChef Rating API!",
            "usage": "To get a user's rating, use the /user/<username> path.",
            "example": "https://<your-site-name>.netlify.app/user/gennady"
        }
        return {
            'statusCode': 200, # Return 200 OK
            'headers': {'Content-Type': 'application/json'},
            'body': json.dumps(welcome_message, indent=2)
        }

    # The rest of the function continues as before if a username is present
    user_data, error = get_codechef_data(username)

    if error:
        return {
            'statusCode': 404,
            'headers': {'Content-Type': 'application/json'},
            'body': json.dumps({'error': error})
        }

    try:
        rating_history = user_data.get("date_versus_rating", {}).get("all", [])
        
        if not rating_history:
            return {
                'statusCode': 200,
                'headers': {'Content-Type': 'application/json'},
                'body': json.dumps({'username': username, 'status': 'Unrated', 'currentRating': 0, 'peakRating': 0})
            }

        latest_contest = rating_history[-1]
        current_rating = int(latest_contest['rating'])
        all_ratings = [int(contest['rating']) for contest in rating_history]
        peak_rating = max(all_ratings)

        response_body = {
            'username': username,
            'status': 'Success',
            'currentRating': current_rating,
            'peakRating': peak_rating,
            'ratingHistory': rating_history
        }

        return {
            'statusCode': 200,
            'headers': {'Content-Type': 'application/json'},
            'body': json.dumps(response_body)
        }
        
    except (KeyError, IndexError, ValueError):
        return {
            'statusCode': 500,
            'headers': {'Content-Type': 'application/json'},
            'body': json.dumps({'error': 'Failed to parse user rating data from page.'})
        }

File: netlify/functions/rating.js
const https = require('https');

async function getCodeChefData(username) {
    return new Promise((resolve, reject) => {
        const url = `https://www.codechef.com/users/${username}`;
        const options = {
            headers: {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
            }
        };

        https.get(url, options, (res) => {
            let data = '';
            
            res.on('data', (chunk) => {
                data += chunk;
            });
            
            res.on('end', () => {
                try {
                    // Extract the JSON data from the page
                    const match = data.match(/jQuery\.extend\(Drupal\.settings, (\{.*\})\);/);
                    if (!match) {
                        reject(new Error('Could not find data object in page source'));
                        return;
                    }
                    
                    const userData = JSON.parse(match[1]);
                    resolve(userData);
                } catch (error) {
                    reject(new Error('Failed to parse user data'));
                }
            });
        }).on('error', (error) => {
            if (error.code === 'ENOTFOUND' || error.message.includes('404')) {
                reject(new Error(`User '${username}' not found or profile is private`));
            } else {
                reject(new Error(`Network error: ${error.message}`));
            }
        });
    });
}

exports.handler = async (event, context) => {
    // Set CORS headers
    const headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Content-Type',
        'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
        'Content-Type': 'application/json'
    };

    // Handle preflight requests
    if (event.httpMethod === 'OPTIONS') {
        return {
            statusCode: 200,
            headers,
            body: ''
        };
    }

    try {
        // Debug: Log the entire event to see what we're receiving
        console.log('Event:', JSON.stringify(event, null, 2));

        // Get username from query parameters or path
        let username = event.queryStringParameters?.username;

        // If no username in query params, try to extract from path
        if (!username && event.path) {
            const pathMatch = event.path.match(/\/rating\/(.+)$/);
            if (pathMatch) {
                username = pathMatch[1];
            }
        }

        // If no username provided, return welcome message
        if (!username) {
            const welcomeMessage = {
                message: "Welcome to the Unofficial CodeChef Rating API!",
                usage: "To get a user's rating, use the /user/<username> path.",
                example: "https://codechefapi.netlify.app/user/gennady"
            };
            
            return {
                statusCode: 200,
                headers,
                body: JSON.stringify(welcomeMessage, null, 2)
            };
        }

        // Fetch user data from CodeChef
        const userData = await getCodeChefData(username);
        
        // Extract rating history
        const ratingHistory = userData?.date_versus_rating?.all || [];
        
        if (ratingHistory.length === 0) {
            return {
                statusCode: 200,
                headers,
                body: JSON.stringify({
                    username: username,
                    status: 'Unrated',
                    currentRating: 0,
                    peakRating: 0
                })
            };
        }

        // Calculate current and peak ratings
        const latestContest = ratingHistory[ratingHistory.length - 1];
        const currentRating = parseInt(latestContest.rating);
        const allRatings = ratingHistory.map(contest => parseInt(contest.rating));
        const peakRating = Math.max(...allRatings);

        const responseBody = {
            username: username,
            status: 'Success',
            currentRating: currentRating,
            peakRating: peakRating,
            ratingHistory: ratingHistory
        };

        return {
            statusCode: 200,
            headers,
            body: JSON.stringify(responseBody)
        };

    } catch (error) {
        console.error('Error:', error.message);
        
        return {
            statusCode: 404,
            headers,
            body: JSON.stringify({
                error: error.message
            })
        };
    }
};
